mesonbuild/build.py

```diff
@@ -3076,3 +3087,97 @@ def save(obj: Build, filename: str) -> None:
             pickle.dump(obj, f)
     finally:
         obj.environment.coredata = cdata
+
+
+@dataclass(eq=False)
+class ResponseFile(HoldableObject):
+
+    args: T.List[CommandArg]
+    subdir: str
+    environment: environment.Environment
+
+    prefix: str = '@'
+    quote: T.Union[bool, str] = True
+    max_args_length: T.Optional[int] = None
+    separator: str = ' '
+    relative_paths: bool = False
+
+    def __post_init__(self):
+        self._response_file: T.Optional[Path] = None
+
+        if self.max_args_length is None:
+            self.max_args_length = get_rsp_threshold()
+
+        self.args = [self.processStringArg(a) for a in self.args]
+
+    def has_response_file(self) -> bool:
+        return self._response_file is not None
+
+    def configure(self, filename: str) -> None:
+        if not self.args:
+            return
+
+        quoted_args = self.separator.join(self.quote_fct(a) for a in self.generateArgList(self.args))
+        if len(quoted_args) <= self.max_args_length:
+            mlog.debug(f'Estimated arg length {len(quoted_args)} <= {self.max_args_length}')
+            return
+
+        self._response_file = Path(self.subdir, filename).with_suffix('.rsp')
+
+        mlog.debug(f'Writing rsp file: {self._response_file}')
+
+        response_file_fullpath = Path(self.environment.build_dir, self._response_file)
+        response_file_fullpath.parent.mkdir(exist_ok=True)
+        response_file_fullpath.write_text(quoted_args, encoding='utf-8')
+
+    def generateArgList(self, args: T.List[CommandArg]) -> Generator[FileOrString, None, None]:
+        for arg in args:
+            if isinstance(arg, (BuildTarget, CustomTarget, CustomTargetIndex)):
+                yield File(True, arg.subdir, arg.get_filename())
+            else:
+                yield arg
+
+    def processStringArg(self, arg: CommandArg) -> CommandArg:
+        """ Convert string to File if it is an existing source file """
+        if isinstance(arg, str):
+            if Path(self.environment.source_dir, self.subdir, arg).exists():
+                return File(False, self.subdir, arg)
+        return arg
+
+    def get_outputs(self) -> T.List[CommandArg]:
+        if self._response_file:
+            return [self.prefix + str(self._response_file)]
+
+        return self.args
+
+    def get_dependencies(self) -> Generator[T.Union[BuildTarget, CustomTarget], None, None]:
+        if self._response_file:
+            for arg in self.args:
+                if isinstance(arg, (BuildTarget, CustomTarget)):
+                    yield arg
+                elif isinstance(arg, CustomTargetIndex):
+                    yield arg.target
+
+    def get_depend_files(self) -> Generator[File, None, None]:
+        if self._response_file:
+            for arg in self.args:
+                if isinstance(arg, File):
+                    yield arg
+
+    def quote_fct(self, value: FileOrString) -> str:
```

Comment:
Rename this function to `quote()`. You are not returning a function here so the `fct` suffix makes no sense.

Reply:
Maybe rename it to `quote_arg` then

Reply:
then it would clash with `mesonlib.quote_arg`...

Reply:
Ok. Use a fun name like `quote_me()`? Or the underscore variant as you said.